File: openpifpaf/decoder/utils/occupancy.py
import logging
import numpy as np

from ...functional import scalar_nonzero_clipped_with_reduction

LOG = logging.getLogger(__name__)


def scalar_square_set(field, x, y, sigma):
    minx = max(0, int(x - sigma))
    miny = max(0, int(y - sigma))
    # +1: for non-inclusive boundary
    # There is __not__ another plus one for rounding up:
    # The query in occupancy does not round to nearest integer but only
    # rounds down.
    maxx = max(minx + 1, min(field.shape[1], int(x + sigma) + 1))
    maxy = max(miny + 1, min(field.shape[0], int(y + sigma) + 1))
    field[miny:maxy, minx:maxx] = 1


class Occupancy():
    def __init__(self, shape, reduction, *, min_scale=None):
        assert len(shape) == 3
        if min_scale is None:
            min_scale = reduction
        assert min_scale >= reduction

        self.reduction = reduction
        self.min_scale_reduced = min_scale / reduction

        self.occupancy = np.zeros((
            shape[0],
            int(shape[1] / reduction) + 1,
            int(shape[2] / reduction) + 1,
        ), dtype=np.uint8)
        LOG.debug('shape = %s, min_scale = %d', self.occupancy.shape, self.min_scale_reduced)

    def __len__(self):
        return len(self.occupancy)

    def set(self, f, x, y, sigma):
        """Setting needs to be centered at the rounded (x, y)."""
        if f >= len(self.occupancy):
            return

        xi = x / self.reduction
        yi = y / self.reduction
        si = max(self.min_scale_reduced, sigma / self.reduction)
        scalar_square_set(self.occupancy[f], xi, yi, si)

    def get(self, f, x, y):
        """Getting needs to be done at the floor of (x, y)."""
        if f >= len(self.occupancy):
            return 1.0

        # floor is done in scalar_nonzero_clipped below
        return scalar_nonzero_clipped_with_reduction(self.occupancy[f], x, y, self.reduction)

File: cpp/src/occupancy.cpp
#include <algorithm>
#include <math.h>

#include "openpifpaf/decoder/utils/occupancy.hpp"


namespace openpifpaf {
namespace decoder {
namespace utils {


void Occupancy::set(int64_t f, double x, double y, double sigma) {
    if (reduction != 1.0) {
        x /= reduction;
        y /= reduction;
        sigma = fmax(min_scale_reduced, sigma / reduction);
    }

    auto minx = std::clamp(int64_t(x - sigma), int64_t(0), occupancy.size(2) - 1);
    auto miny = std::clamp(int64_t(y - sigma), int64_t(0), occupancy.size(1) - 1);
    // # +1: for non-inclusive boundary
    // # There is __not__ another plus one for rounding up:
    // # The query in occupancy does not round to nearest integer but only
    // # rounds down.
    auto maxx = std::clamp(int64_t(x + sigma), minx + 1, occupancy.size(2));
    auto maxy = std::clamp(int64_t(y + sigma), miny + 1, occupancy.size(1));
    occupancy.index_put_({f, at::indexing::Slice(miny, maxy), at::indexing::Slice(minx, maxx)}, 1);
}


bool Occupancy::get(int64_t f, double x, double y) {
    if (f >= occupancy.size(0)) return 1;

    if (reduction != 1.0) {
        x /= reduction;
        y /= reduction;
    }

    auto xi = std::clamp(int64_t(x), int64_t(0), occupancy.size(1) - 1);
    auto yi = std::clamp(int64_t(y), int64_t(0), occupancy.size(0) - 1);
    return occupancy.index({yi, xi}).item<uint8_t>() != 0;
}


} // namespace utils
} // namespace decoder
} // namespace openpifpaf
